## Exploratory Modelling and Feature Analysis

This part is going to investigate the transformed electricity dataset to identify key features, 
patterns, or states that may inform the final model, or reveal changes that need to be addressed
in the main pipeline. 

In [1]:
import pandas as pd

df = pd.read_parquet("../data/processed/electricity_transformed.parquet")

KeyboardInterrupt: 

### Inital Exploration

In [3]:
df.shape

(52800, 44)

In [4]:
df.head

<bound method NDFrame.head of                       timestamp  curtailment  curtailment_solar_utility  \
0     2020-09-15 00:00:00+10:00          0.0                        0.0   
1     2020-09-15 01:00:00+10:00          0.0                        0.0   
2     2020-09-15 02:00:00+10:00          0.0                        0.0   
3     2020-09-15 03:00:00+10:00          0.0                        0.0   
4     2020-09-15 04:00:00+10:00          0.0                        0.0   
...                         ...          ...                        ...   
52795 2026-09-23 19:00:00+10:00          0.0                        0.0   
52796 2026-09-23 20:00:00+10:00          0.0                        0.0   
52797 2026-09-23 21:00:00+10:00          0.0                        0.0   
52798 2026-09-23 22:00:00+10:00          0.0                        0.0   
52799 2026-09-23 23:00:00+10:00          0.0                        0.0   

       curtailment_wind       demand  demand_gross  flow_exports  \
0

In [5]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 52800 entries, 0 to 52799
Data columns (total 44 columns):
 #   Column                                 Non-Null Count  Dtype                    
---  ------                                 --------------  -----                    
 0   timestamp                              52800 non-null  datetime64[us, UTC+10:00]
 1   curtailment                            52800 non-null  float64                  
 2   curtailment_solar_utility              52800 non-null  float64                  
 3   curtailment_wind                       52800 non-null  float64                  
 4   demand                                 52800 non-null  float64                  
 5   demand_gross                           52800 non-null  float64                  
 6   flow_exports                           52800 non-null  float64                  
 7   flow_imports                           52800 non-null  float64                  
 8   generation_renewable                 

In [6]:
df.isna().sum()

timestamp                                0
curtailment                              0
curtailment_solar_utility                0
curtailment_wind                         0
demand                                   0
demand_gross                             0
flow_exports                             0
flow_imports                             0
generation_renewable                     0
generation_renewable_with_storage        0
price                                    0
renewable_proportion                     0
renewable_with_storage_proportion        0
emissions                                0
power                                    0
was_imputed                              0
curtailment_negative                     0
hour_sin                                 0
hour_cos                                 0
day_of_week_sin                          0
day_of_week_cos                          0
month_sin                                0
month_cos                                0
is_weekend 

In [7]:
df.columns.tolist()

['timestamp',
 'curtailment',
 'curtailment_solar_utility',
 'curtailment_wind',
 'demand',
 'demand_gross',
 'flow_exports',
 'flow_imports',
 'generation_renewable',
 'generation_renewable_with_storage',
 'price',
 'renewable_proportion',
 'renewable_with_storage_proportion',
 'emissions',
 'power',
 'was_imputed',
 'curtailment_negative',
 'hour_sin',
 'hour_cos',
 'day_of_week_sin',
 'day_of_week_cos',
 'month_sin',
 'month_cos',
 'is_weekend',
 'price_lag_1h',
 'price_lag_24h',
 'price_lag_168h',
 'price_rolling_mean_24h',
 'price_rolling_std_24h',
 'demand_lag_1h',
 'demand_lag_24h',
 'demand_lag_168h',
 'demand_rolling_mean_24h',
 'demand_rolling_std_24h',
 'generation_renewable_lag_1h',
 'generation_renewable_lag_24h',
 'generation_renewable_lag_168h',
 'generation_renewable_rolling_mean_24h',
 'generation_renewable_rolling_std_24h',
 'curtailment_lag_1h',
 'curtailment_lag_24h',
 'curtailment_lag_168h',
 'curtailment_rolling_mean_24h',
 'curtailment_rolling_std_24h']

### Seperating Feature Groups

In [8]:
# Core features
core_features = [
    "curtailment",
    "curtailment_solar_utility",
    "curtailment_wind",
    "demand",
    "demand_gross",
    "flow_exports",
    "flow_imports",
    "generation_renewable",
    "generation_renewable_with_storage",
    "price",
    "renewable_proportion",
    "renewable_with_storage_proportion",
    "emissions",
    "power"
]

# Time-based features
time_features = [
    "hour_sin",
    "hour_cos",
    "day_of_week_sin",
    "day_of_week_cos",
    "month_sin",
    "month_cos",
    "is_weekend"
]

# Data quality features
quality_flags = [
    "was_imputed",
    "curtailment_negative"
]

# Lagged features
lag_features = [
    "price_lag_1h",
    "price_lag_24h",
    "price_lag_168h",
    "demand_lag_1h",
    "demand_lag_24h",
    "demand_lag_168h",
    "generation_renewable_lag_1h",
    "generation_renewable_lag_24h",
    "generation_renewable_lag_168h",
    "curtailment_lag_1h",
    "curtailment_lag_24h",
    "curtailment_lag_168h"
]

# Rolling features
rolling_features = [
    "price_rolling_mean_24h",
    "price_rolling_std_24h",
    "demand_rolling_mean_24h",
    "demand_rolling_std_24h",
    "generation_renewable_rolling_mean_24h",
    "generation_renewable_rolling_std_24h",
    "curtailment_rolling_mean_24h",
    "curtailment_rolling_std_24h"
]

timestamp_col = "timestamp"

### Investigating Core Features

In [9]:
X = df[core_features].copy()
X.describe().T

,count,mean,std,min,25%,50%,75%,max
curtailment,52800.0,106.329910,278.784519,-65.397000,0.000000,0.033100,70.840017,3503.327600
curtailment_solar_utility,52800.0,93.339836,249.393985,-65.397000,0.000000,0.000000,50.607098,3318.139467
curtailment_wind,52800.0,12.990074,51.534974,-25.700983,0.000000,0.000000,0.389833,1314.630875
demand,52800.0,7524.840933,1365.634929,2577.155833,6589.504166,7399.240833,8292.109375,13641.368333
demand_gross,52800.0,8613.144084,1516.777145,5411.577608,7455.836335,8550.551850,9573.128812,16134.240858
flow_exports,52800.0,151.195615,207.319850,0.000000,0.000000,47.823967,247.912883,1546.858323
flow_imports,52800.0,745.377861,411.946338,0.000000,448.290304,736.492139,1020.767551,2522.498118
generation_renewable,52800.0,2601.494633,2028.636065,26.355308,1065.931323,1786.491033,3807.208209,10898.422467
generation_renewable_with_storage,52800.0,2742.933349,2071.721107,26.355308,1176.059590,1967.275779,3911.019604,11477.422475
price,52800.0,109.136487,307.248261,-328.295833,45.723436,76.186250,117.703958,16157.503333


The variation in scales highlights the neede for scaling. From this we can also tell:
- cultailment, curtailment_solar_utility, and curtailment_wind are right skewed 
- price indicates major pricce spikes with the max of 16158
- demand is more stable

These high observations are not neccessarily outliers, but rather indicators of extreme operating conditions.

### Feature Relationships

In [1]:
correlation_matrix = X.corr()

correlation_matrix.round(2)

import matplotlib.pyplot as plt

plt.figure(figsize=(12, 10))
plt.imshow(correlation_matrix)

plt.colorbar(label="correlation")
plt.xticks(
    range(len(correlation_matrix.columns)),
    correlation_matrix.columns,
    rotation=90
)
plt.yticks(
    range(len(correlation_matrix.columns)),
    correlation_matrix.columns
)

plt.title("Core Electricity Features")
plt.tight_layout()
plt.show()

KeyboardInterrupt: 